### ISSUE TITLE
toNamedCsv

### ISSUE THREAD

**Issue by nevrome** [domdere/cassava-conduit#34]
Great library, thanks for developing and maintaining it!

One weak point, though, is writing .csv files back to the file system. Probably you didn't need this functionality for your use-case. To my understanding at the moment cassava-conduit only includes a very basic `toCsv` function:

```Haskell
toCsv :: (Monad m, ToRecord a) => EncodeOptions -> ConduitT a BS.ByteString m ()
toCsv opts = {-# SCC toCsv_p #-} map $ BSL.toStrict . encodeWith opts . pure
```

with a comment indicating that this was more of an afterthought: "Streams from csv to text, does not create headers...".

I would like to know if it would be possible to add a function `toNamedCsv`, which makes use of the `ToNamedRecord` class in cassava to automatically write with column names. This is important functionality for my application. Are there special challenges involved with this or was it just not a priority for you so far?

I saw that there is an alternative with the csv-conduit library, which seems to have this functionality. But of course I would prefer to stay with cassava, which I understand already, over a completely different library.

**Comment by domdere** [comment]
It wasn't really an after thought but an intentional part of the design, a
conduit is a thing that takes in a stream of x's and produces a stream of
y's with fixed memory bounds etc..., so in case it's a stream of input
elements and for each one it produces a serialized  bytestring to pipe into
a file. The header line however is a single line, it doesn't need to be
streamed, it doesn't correspond to a single one of the input elements, and
writing a single line of text to a file can be done in a bounded manner by
trivial means with ordinary strict IO operations. You don't really need a
conduit to do this.

The intended way to do this is to generate ur header line using cassava and
output it to your file handle with BS.hPutStr or similar, and then open up
your conduit and begin streaming your rows into it.

**Comment by nevrome** [comment]
Thank you for the fast reply! What you say makes a lot of sense! I'm struggling a bit with the concrete implementation, though.

I have a lot of nested record types. I assume I can construct the header line you describe with `Data.Csv.headerOrder` for any of them, if I have defined my desired instances for `ToNamedRecord`. But then I lack the necessary instances for `ToRecord` required for `Data.Csv.Conduit.toCsv`, right? If I understand this correctly, then I would have to maintain instances for both `ToNamedRecord` and `ToRecord`, which is a lot of boilerplate and even introduces the possibility that I mess up the order somehow.

Pardon my ignorance. I'm sure I'm missing something here.

**Comment by domdere** [comment]
Sure in the general case in Haskell if you have a typeclass instance and you want to derive an instance of another typeclass from it so that you want to maintain only one instance you could do it be define a `newtype` and defining an instance of `FromRecord` for `Newtype a` that uses the `FromNamedRecord` instance of `a`.

However in this specific case, it probably be more proficient to use generics like in the cassava example [here](https://hackage.haskell.org/package/cassava-0.5.3.0/docs/Data-Csv.html#g:3)

You only maintain the record type for your data type and both the `FromNamedRecord` and `FromRecord` instances are automatically generated from that, its fairly light on boilerplate. If you don't actually control the definition of this type, then write up a separate record type and write a map function that maps the type to the new record type and run them through the conduit...

**Comment by nevrome** [comment]
I took a deeper look into cassava and conduit to come up with a solution. Even had a little help from ChatGPT, which was an interesting experiment.

I manually added instances for the `DefaultOrdered` cassava class to my record types (some of them could be done with generics, but I like to be in control, after all). This enabled me to use the `headerOrder` function to get the correct header string even for sum types.

The biggest issue was writing the header to the file within the stream. I couldn't do that before starting the conduit pipeline, because the specific instance of the type I want to write to a .csv file (across various sum types) would only become clear within the pipeline.

Here's what I came up with:

```Haskell
sinkNamedCSV :: (MonadResource m, Csv.ToRecord a, Csv.DefaultOrdered a) => FilePath -> ConduitT a Void m ()
sinkNamedCSV path =
    Con.bracketP (openFile path WriteMode) hClose $ \handle ->
           writeHeaderCSV handle
        .| ConCsv.toCsv encodingOptions
        .| ConL.mapM_ (liftIO . Bchs.hPutStr handle)
    where
        writeHeaderCSV :: (MonadIO m, Csv.DefaultOrdered i) => Handle -> ConduitT i i m ()
        writeHeaderCSV handle = do
            flagRef <- liftIO $ newIORef True
            ConL.mapM $ \val -> do
                -- run the action only for the first element
                flag <- liftIO $ readIORef flagRef
                when flag $ do
                    liftIO $ BB.hPutBuilder handle $ CsvB.encodeHeaderWith encodingOptions $ Csv.headerOrder val
                    liftIO $ modifyIORef flagRef not
                -- continue processing the rest of the stream
                return val
```

This seems to work. No obvious memory leaks or performance losses and the output file looks as expected.

The only thing that annoys me a bit is the way I apply the action to only the first element of the stream. I feel there must me a shorter and more idiomatic way to do this. Maybe you have a suggestion.

Otherwise you can close this issue. Again: Thank you very much for your help :+1:

**Comment by domdere** [comment]
Its also generally just really strange to be pumping sum types into a CSV
conduit... a CSV record row is by it's very nature a record and hence more
suited to record types, you would either be discarding some "invalid"
values, or all values can in essence be mapped to a custom "CSV" record
type like the Person type in the example, in which case you map all the
values to that type and then pump them into the conduit pipeline with your
generic derived instances (I'm not really sure I get your aversion to the
generic instances, you want full control that comes with writing instances
by hand but want to avoid boilerplate, those two goals seem to be at odd
ends, and you technically do have control, if circumstances change and the
generic instances no longer suit you, you have the freedom/control to
remove the generic instances and define your desired instances by hand).

But anyway, to be using a sum type in this instance where not all values of
the type have the same CSV header, your type really doesn't represent a CSV
record at all, and to pipe them into a CSV stream is to introduce an
heuristic, a model that doesn't line up with the reality, and they will
always compicate your programs .

**Comment by nevrome** [comment]
Haha - alright. You may have a point here. It's indeed often a software design problem, when you run into such weird edge cases.

To some degree the header is of course known from the beginning based on the arguments submitted by the user. But this is complex logic that I would rather like to implement once in my actual processing pipeline, and not a second time independently for the header. If I can solve it on a type level, then I probably should, right?

For example my command line software (part of a research project, can't make the code public yet) takes spatial coordinates either in a Cartesian coordinate system (x and y) or as longitudes and latitudes on a sphere. This is a big difference, for example regarding distance calculation. I represent this in a type that looks something like this:

```Haskell
data SpatialPosition = SpatialPositionCartesian CartesianPosition | SpatialPositionLongLat LongLatPosition
```

In the header of a .csv file generated from data in these types I want either `x` and `y` or `long` and `lat`.

Also my pipeline uses different algorithms to calculate some measure of probability. Depending on the algorithm, the output types may look slightly different, which again should be represented by the columns in the output .csv.

For me it makes sense to implement this system as described. But as I tend to overlook obvious solutions I'm eager to hear what you think! :smile:

Regarding the generics: To my understanding generics use the record field names for column names. So they have to follow naming conventions of Haskell functions and must not overlap within my types module. I would like to be a bit a bit more flexible regarding the column names (even beyond `fieldLabelModifier`).
